### ISSUE TITLE
YOLO best.pt model exports to  onnx graph with negative axes in nodes

### ISSUE THREAD

**Issue by ShonSat** [ShonSat/tiger_watch#6]
Issue: PyTorch model trained on YOLO pytorch framework  has ONNX graph has nodes with negative axis. 
Negative axis in nodes are not supported on older Jetson Xavier 
TensorRT: v8.5.2
SW: https://developer.nvidia.com/embedded/jetpack-sdk-516

Original issue: https://github.com/ShonSat/tiger_watch/issues/3   
> ###  Issue:   ONNX-to-TensorRT conversion fails. 
> Failing CMD: trtexec --onnx=best.onnx --saveEngine=best_model.engine --fp16
> 
> There are 2 parts that I need to look at.
> ### Part 1. Both best.pt (pytorch model) and best.onnx (exported from best.pt with Ultralitics and pytorch) contain  INT64 data types  and Jetson HW is limited to INT32 data types for inference.
> During ONNX-to-TensorRT conversion,  trtexec sees INT64 operators in ONNX graph and attempts to Downcast  data type to INT32, but fails. 
> Even  if  conversion  to TensorRT ( .engine format) succeeds,  resulting TRT engine's INT64 values may exceed  INT32 limits  (overflow ) => broken model accuracy
> Ref: https://forums.developer.nvidia.com/t/onnx-model-int64-weights/124248; 
> 
>  PyTorch operations like shape or range default to INT64. best.onnx graph/blueprint uses INT64 data types for indices, shape counters, and tensor dimension operations (such as reshaping or flattening layers),
>  but  Jetson HW supports  INT32 data types for inference.
>  
> In this project  I have to look at 2 steps prior to ONNX to TRT conversion:
>       a)      training:  tiger_watch/train_Ultralytics_YOLO_model.py   that spits out best.pt
>       b)     export YOLO model  (best.pt) to ONNX (best.onnx) format: tiger_watch/export_YOLO_to_ONNX.py
> 
> 
> #### a)  YOLO training script and closer look at  PyTorch weight files:  
> --model = yolo26n.pt   
> --imgsz = 640  
> 
> Training image size 640 pixels should easily fit in INT32 data types (so no overflow in TRT engine downstream)
> 
> ##### model weights and data types in .pt  | PyTorch weights
> Ultralytics YOLO26 Nano (yolo26n.pt) base model  and my custom best.pt model are trained using PyTorch and  actual weight parameters (convolutions, linear layers, etc.) that learn how to detect tigers are stored as floating-point numbers usually 32-bit floats (FP32) or 16-bit floats (FP16)
> 
> ##### tensor shapes and grid dimensions | torch.int64 (torch.long)
> Pytorch by default uses torch.int64 (torch.long) for matrix shapes, slicing indices, and anchor grids.
> https://pytorchcourse.com/01-tensors/03_data_types_and_devices/
> 
> ##### metadata of the .pt file 
> .pt file preserves metadata like the current epoch count, class IDs, and optimizer step counters. Python and PyTorch automatically default these whole numbers to 64-bit integers
> 
>         
> #### b) ONNX  (Open Neural Network Exchange | https://onnx.ai)  and datatypes
> best.onnx is a compiled version of best.pt  and represents best.pt as computational graph 
> 
> best.onnx graph/blueprint uses INT64 data types for indices, shape counters, and tensor dimension operations (such as reshaping or flattening layers)
> 
> https://docs.ultralytics.com/modes/export
> 
> Info about ONNX datatypes: https://onnx.ai/onnx/intro/concepts.html#supported-types 
> "ONNX is strongly typed and its definition does not support implicit cast. ONNX does not allow addition of two tensors or matrices with different types, even if other languages do. That’s why an explicit cast must be inserted in a graph." 
> 
> During export from .pt to onnx format,  pytorch maps out math operations into a universal computational graph. 
> https://docs.pytorch.org/docs/main/onnx_export.html 
> 
> ### Part 2:  best.onnx graph has  nodes with  negative axis.
> trtexec fails with trace pointing to this part in onnx graph : 
> ```
> name: "/model.23/TopK"
> op_type: "TopK"
> attribute {
> name: "axis"
> i: -1
> type: INT
> } 
> ....
> ERROR: onnx2trt_utils.cpp:345 In function convertAxis:
> [8] Assertion failed: (axis >= 0 && axis <= nbDims) && "Axis must be in the range [0, nbDims]."
> ```
> TopK layers are closer to end of the graph.
> 
> Why ONNX accepts and uses negative value for axis, but TensorRT does not?  

 _Originally posted by @ShonSat in [#3](https://github.com/ShonSat/tiger_watch/issues/3#issuecomment-5233478072)_

**Comment by ShonSat** [comment]
Latest TensorRT version supports negative axes, but old TensorRT v8.5 on older jetsons are limited to only positive 
axes.  https://forums.developer.nvidia.com/t/onnx-tensorrt-convertaxis-assertion-failed/113028/14

// Install onnx-simplifier
pip install onnx-simplifier

// Simplify and fix the ONNX model
python -m onnxsim best.onnx best_simplified.onnx

<img width="840" height="851" alt="Image" src="https://github.com/user-attachments/assets/782f6059-af22-4e05-a11e-04b4e6a3fd9a" />

**Comment by ShonSat** [comment]
Attempt to export with int8 quantization  https://docs.ultralytics.com/modes/export#quantization-options also failed to remove negative axes.  
model.export(format='onnx', imgsz=640, dynamic=False, simplify=False, quantize=8, verbose=True, opset=11, data='/home/shon/Sandbox/datasets/YOLO_wildlife/')

**Comment by ShonSat** [comment]
I'm tweaking parameters for export to ONNX format  (export_OLO_to_ONNX.py)  to address https://github.com/ShonSat/tiger_watch/issues/6

simplify=False  # Ultralytics uses a graph simplifier (onnxslim or onnx-simplifier) by default to merge layers.
                  Sometimes this process collapses explicit axis mappings into negative values.
dynamic=False  # is critical to set dynamic=False, because TensorRT likes fixed image shapes for optimization
opset=12       # TensorRT 8.5.2 on Jetson supports onnx opset 17 or less.
quantize=8     # quantization step requires path to the training dataset yaml
data='/home/shon/Sandbox/datasets/YOLO_wildlife/',  # needed for quantization
nms=True       # NMS wraps TopK in a TensorRT-compatible plugin and model runs on GPU.
max_det=100,   #topk: max number of detections, to reduce complexity
conf=0.25,     # confidence for NMS
iou=0.45,      # IOU threshold for NMS
 
 
`
// load best custom-trained weights from the run directory
model = YOLO("runs/train/tiger_watch_yolo/weights/best.pt")
//model = YOLO("runs/detect/train/tiger_watch_yolo_v9s/weights/best.pt")

// Set the model to end-to-end mode BEFORE exporting
model.model.end2end = True

// Export to ONNX
model.export(format='onnx',
             imgsz=640,
             dynamic=False,
             simplify=True,
             verbose=True,
             opset=12,
             nms=True,      # Add NMS operation with its arguments to the graph.
             max_det=100,   # topk: max number of detections
             conf=0.25,     # confidence for NMS
             iou=0.45,      # IOU threshold for NMS
)

`

Resulting onnx files are visibly different by filesize and parameters number :

<img width="716" height="438" alt="Image" src="https://github.com/user-attachments/assets/0d57dcfb-d37b-4ead-a73f-7dead613d27f" />


Next step to convert each of them to TensorRT engine format and see which parameter variation passes.

**Comment by ShonSat** [comment]
See notes in:  https://github.com/ShonSat/tiger_watch/blob/main/export_YOLO_to_ONNX.py

Minimum arguments needed to export pytorch model to onnx that is acceptable for TensorRT 8.5.2 on Jetson should include 
opset=17 or less, use stable  version opset=12:

```
model.export(format='onnx', imgsz=640, dynamic=False, verbose=True, opset=12)

```
Created onnx files with various parameters, and converted each to TRT (engine) format: run convert_ONNX_to_TensorRT.sh on jetson

[onnx files and corresponding engine files on conversion success:](https://github.com/ShonSat/tiger_watch/tree/main/runs/train/tiger_watch_yolo/weights)

[trtexec logs:](https://github.com/ShonSat/tiger_watch/tree/main/logs)
 
Failure reported on this bug reproduced when I removed opset argument from model.export function:
https://github.com/ShonSat/tiger_watch/blob/main/logs/convert_ONNX_to_TensorRT.sh_best_dynamicFalse.onnx.log 
https://github.com/ShonSat/tiger_watch/blob/main/logs/convert_ONNX_to_TensorRT.sh_best_dynamicOff_end2endOn.onnx.log
